cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/AppMetadataStore.java

```diff
@@ -501,11 +501,16 @@ private boolean validateExistingRecords(List<RunRecordMeta> existingRecords, Pro
                  "Skip recording program {}.", existingRecords, programId, pid, recordType);
       return false;
     }
-    if (existingRecords.isEmpty()) {
-      // If existing record is not allowed, only empty existing records is valid
-      if (allowedStatuses.isEmpty() && allowedWithLogStatuses.isEmpty()) {
+    // If existing record is not allowed, only empty existing records is valid
+    if (allowedStatuses.isEmpty() && allowedWithLogStatuses.isEmpty()) {
+      if (existingRecords.isEmpty()) {
         return true;
       }
+      LOG.error("No existing run record meta should exist for program '{}' pid '{}' but found '{}'.",
+                programId, pid, recordType, existingRecords);
```

Comment:
I think the messages need some work, so that our users can also understand them. Let's do that in a follow-up PR.